File: 116-populating-next-right-pointers-in-each-node/116-populating-next-right-pointers-in-each-node.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""
from collections import deque
class Solution:
    def connect(self, root: 'Node') -> 'Node':
        cur = root

        flag = True
        while(cur):
            #set the next
            if flag == True:
                if cur.left:
                    nxt = cur.left
                else:
                    nxt = cur.right
                flag = False
            #find the next real pointer
            p = cur.next
            if not p:
                if cur.left:
                    cur.left.next = cur.right
                    cur.right.next = None
                    cur = nxt
                    flag = True
                    continue
                else:
                    break
            if cur.left:
                cur.left.next = cur.right
                cur.right.next = p.left
                cur = p
            else:
                break
        return root
        
        

File: 116-populating-next-right-pointers-in-each-node/116-populating-next-right-pointers-in-each-node.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {}
    
    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, Node _left, Node _right, Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
};
*/

class Solution {
    public Node connect(Node root) {
        //@input: a perfect binary tree
        //@output: the head with next pointer 
        //bfs, start from the leftmost node in the level one up.
        //then set the next pointer of the node's left children to the node's right children
        //set the next pointer of the node's right children to the next node's left
        //@edge case: the root is null
        
        Node prev = root;
        while(prev != null){
            connectNextLevel(prev);
            prev = prev.left;
        }
        return root;
    }
    public void connectNextLevel(Node root){
        Node cur = root;
        while(cur != null && cur.left != null){
            cur.left.next = cur.right;
            if(cur.next != null){
                cur.right.next = cur.next.left;
            }
            cur = cur.next;
        }
    }
}